doc/neps/nep-XXXX-SIMD-optimizations.rst

```diff
@@ -0,0 +1,249 @@
+  a single ufunc will be expressed as a set of loops, going from a minimal to
+  a maximal set of possibly availabe intrinsics.
+- At compile time, compiler macros and CPU detection are used to turn the
+  abstract intrinsics into concrete intrinsic calls. Any intrinsics not
+  available on the platform, either because the CPU does not support them
+  (and so cannot be tested) or because the abstract intrinsic does not have a
+  parallel concrete intrinsic on the platform will not error, rather the
+  corresponding loop will not be produced and added to the set of
+  possibilities.
+- At runtime, the CPU detection code will further limit the set of loops
+  available, and the optimal one will be chosen for the ufunc.
+
+The current NEP proposes only to use the runtime feature detection and optimal
+loop selection mechanism for ufuncs. Future NEPS may propose other uses for the
+proposed solution.
+
+Usage and Impact
+----------------
+
+The end user will be able to get a list of intrinsics available for their
+platform and compiler. Optionally,
+the user may be able to specify which of the loops available at runtime will be
+used, perhaps via an environment variable to enable benchmarking the impact of
+the different loops. There should be no direct impact to naive end users, the
+results of all the loops should be identical to within a small number (1-3)
+ULPs. On the other hand, users with more powerful machines should notice a
+performance boost.
+
+Binary releases - wheels on PyPI and conda packages
+```````````````````````````````````````````````````
+
+The binaries released by this process will be larger since they include all
+possible loops for the architecture. Some packagers may prefer to limit the
+number of loops in order to limit the size of the binaries, we would hope they
+would still support a wide range of families of architectures. Note this
+problem already exists in the Intel MKL offering.
+
+Source builds
+`````````````
+TBD
+- Setting the baseline and set of runtime-dispatchable ISA extensions
+- Behavior when compiler or hardware doesn't support a requested ISA extension
+
+
+How to run benchmarks to assess performance benefits
+````````````````````````````````````````````````````
+
+Adding more code which use intrinsics will make the code harder to maintain.
+Therefore, such code should only be added if it yields a significant
+performance benefit. Assessing this performance benefit can be nontrivial.
+To aid with this, the implementation for this NEP will add a way to select
+which instruction sets can be used at *runtime* via environment variables.
+(name TBD).
+
+
+Diagnostics
+```````````
+
+A new dictionary `__cpu_features__` will be available to python. The keys are
+the available features, the value is a boolean whether the feature is available
+or not. Various new private
+C functions will be used internally to query available features. These
+might be exposed via specific c-extension modules for testing.
```

Comment:
Question: How is the current vision for UFunc dispatching, I can see two main things:

1. The exported function pointers are set on startup to point to the best version.
2. The inner-loop queries the features and set the maximum version (and possible caches it in a `static`).

The current method of just inserting a faster loop is not really nice. My gut feeling would be that 1. is the best option, I do not think we have to be able to change this at run-time.

Reply:
I think this is orthogonal to this NEP, and may be worthy of its own NEP or at least a separate PR.

Reply:
OK, so this NEP limits itself to:
1. We need runtime detection,
2. We want to use the idea of universal SIMD instructions to attack the problem of writing architecture specific loops?

Is the way ufuncs dispatch this complex enough for its own NEP? I assume the ufunc machinery (except the code generation) does not notice that this is happening?

Reply:
I think the whole idea of loop selection based on CPU features is piggy-backing on the dtype loop selection machinery, which was not designed from the ground up for this possibility but does not seem too hacky.

Reply:
I assumed this effort removed that piggy-backing. It is currently not too bad, I agree. However, I do not think it is sustainable really in the case where we have 3 different floating point loops for each loop type. I should probably dig into the code, I assume the proposed solution inserts only the best loop into the ufuncs at startup?

Reply:
the analysis is done at import. Why wouldn't it be sustainable?

Reply:
No, at import time is perfect. having many more loops and picking the first match is what I would think is not nice. But I do not think that is the design here (it is the current one)

Reply:
This NEP does not propose any changes to the current loop selection mechanism, it assumes the current one is good enough.